File: zanompc/viewkey.go
package zanompc

import (
	"fmt"

	"filippo.io/edwards25519"
	"github.com/KarpelesLab/zanolib/zanocrypto"
)

// Domain-separation tags for the deterministic threshold view key. The base
// point is bound to the wallet's spend public key, and both the point and the
// final scalar are tagged so the derived value can never collide with — or be
// linked to — a real on-chain key image (whose base is Hp(stealth_address),
// with no prefix).
var (
	viewKeyBaseDomain   = []byte("ZANO_THRESHOLD_VIEWKEY\x00")
	viewKeySecretDomain = []byte("ZANO_THRESHOLD_VIEWKEY_SECRET\x00")
)

// viewKeyMsg carries one party's contribution w_i*P to the deterministic view
// key derivation.
type viewKeyMsg struct {
	Partial []byte `json:"partial"` // w_i * P
}

// ViewKeyBase returns the fixed base point P = Hp("ZANO_THRESHOLD_VIEWKEY\0" ||
// spendPub) used to derive the deterministic threshold view secret. It is a
// pure function of the group spend public key, so every committee member
// computes the same P.
func ViewKeyBase(spendPub *edwards25519.Point) *edwards25519.Point {
	return zanocrypto.Hp(append(append([]byte{}, viewKeyBaseDomain...), spendPub.Bytes()...))
}

// DeriveViewSecret derives the wallet's view secret key deterministically from
// the shared spend secret x, without anyone reconstructing x. It computes the
// "key image" V = x*P of the fixed base point P = ViewKeyBase(spendPub) — a
// deterministic PRF on x — by summing every party's partial w_i*P over the
// broker, then hashes V to a scalar (the same unclamped Keccak-to-scalar
// convention Zano uses for view keys):
//
//	view_secret = HashToScalar("ZANO_THRESHOLD_VIEWKEY_SECRET\0" || V)
//
// The result is identical on every committee member and reproducible from the
// same key shares, so a threshold wallet has a stable view key (and thus a
// stable address) without a random value or out-of-band agreement. The
// corresponding view public key is view_secret*G.
//
// Call this once per signer; it uses its own broker topic and does not touch
// the KeyImage/SignCLSAG input pairing.
func (s *ThresholdInputSigner) DeriveViewSecret() (*edwards25519.Scalar, error) {
	P := ViewKeyBase(s.spendPub)

	myPartial, err := PartialKeyImage(s.subset, s.subset.Ks, s.params.PartyID().Index, P)
	if err != nil {
		return nil, err
	}

	collected, err := exchange(s.ctx, s.params, "zano:viewkey", &viewKeyMsg{Partial: myPartial.Bytes()})
	if err != nil {
		return nil, err
	}

	partials := make([]*edwards25519.Point, len(collected))
	for i, m := range collected {
		if partials[i], err = pt(m.Partial); err != nil {
			return nil, fmt.Errorf("zanompc: bad view-key partial from party %d: %w", i, err)
		}
	}

	// V = sum_j(w_j*P) = x*P
	V := CombinePoints(partials)

	return zanocrypto.HashToScalar(append(append([]byte{}, viewKeySecretDomain...), V.Bytes()...)), nil
}

File: src/mpc/viewkey.rs
//! A deterministic view key for a threshold wallet.

use super::signer::ThresholdInputSigner;
use crate::crypto::{Point, Scalar};
use crate::error::Result;
use purecrypto::hash::HashAlgorithm;

/// The key-image identifier the view secret is derived under.
///
/// Distinct identifiers give unlinkable secrets from the same shares, so this
/// label is what separates the view key from any other value a committee might
/// derive with the same ceremony.
pub const VIEW_KEY_IDENTIFIER: &[u8] = b"ZANO_THRESHOLD_VIEWKEY";

/// Keccak-256, matching the hash Zano uses everywhere else.
const VIEW_KEY_HASH: HashAlgorithm = HashAlgorithm::Keccak256;

impl ThresholdInputSigner {
    /// Derives the wallet's view secret key deterministically from the shared
    /// spend secret `x`, without anyone reconstructing `x`.
    ///
    /// Zano's usual `view = keccak(spend_secret)` cannot run under MPC, so the
    /// committee instead runs tsslib's threshold key-image ceremony — a
    /// distributed PRF — over [`VIEW_KEY_IDENTIFIER`]: each party contributes
    /// `W_i = λ_i·s_i·P` for a point `P` bound to the identifier and the group
    /// public key, the partials sum to `V = x·P`, and the secret is
    /// `HashToScalar(identifier, V)`. The shared secret never assembles, every
    /// party learns the identical value, and each partial carries a DLEQ proof
    /// so a wrong contribution is caught and its sender named.
    ///
    /// Partials travel point-to-point, never broadcast: `t+1` of them
    /// reconstruct `V`, so the ceremony depends on the per-recipient
    /// confidentiality the broker contract requires.
    ///
    /// The result is stable across runs and across any qualifying committee, so
    /// a threshold wallet has a fixed view key — and therefore a fixed address —
    /// with no random value and no out-of-band agreement. The matching view
    /// public key is `view_secret*G`; see [`ThresholdInputSigner::derive_view_keys`].
    ///
    /// Every committee member must call this, and it must not run concurrently
    /// with a signing session on the same broker.
    pub fn derive_view_secret(&self) -> Result<Scalar> {
        let party = self
            .subset()
            .new_key_image(
                VIEW_KEY_IDENTIFIER.to_vec(),
                self.params().clone(),
                VIEW_KEY_HASH,
            )
            .map_err(|e| crate::err!("zanompc: view key ceremony: {e}"))?;
        let res = party
            .wait()
            .map_err(|e| crate::err!("zanompc: view key ceremony: {e}"))?;
        // Note: `res.public_key` is the *child spend* key (group_pub + secret*G)
        // the ceremony defines for tweaked signing — not the Zano view key.
        Ok(res.secret)
    }

    /// [`ThresholdInputSigner::derive_view_secret`] plus the matching view
    /// public key `view_secret*G`, which is what goes into the address.
    pub fn derive_view_keys(&self) -> Result<(Scalar, Point)> {
        let secret = self.derive_view_secret()?;
        let public = Point::mul_base(&secret);
        Ok((secret, public))
    }
}
